[mlir][spirv] Allow yielding values from loop regions

IgWod-IMG opened the pull request:
This change extends `spirv.mlir.loop` so it can yield values, the same as `spirv.mlir.selection`.

IgWod committed:
[mlir][spirv] Allow yielding values from loop regions

This change extends `spirv.mlir.loop` so it can yield values, the
same as `spirv.mlir.selection`.

kuhar reviewed (approved):
LGTM

IgWod-IMG commented:
Ping @andfau-amd (but no pressure, I just came back from holiday, so I am pinging open PRs :))

andfau-amd commented:
Oh, thanks for the ping, I either missed this or forgot about it. Will take a look now!

andfau-amd commented on mlir/lib/Target/SPIRV/Deserialization/Deserializer.cpp:
```diff
@@ -2003,7 +2003,8 @@ LogicalResult ControlFlowStructurizer::structurize() {
   // arguments will be yielded by the selection region. We do not update uses or
   // erase original block arguments yet. It will be done later in the code.
-  if (!isLoop) {
+  // We do not currently support block arguments in loop merge blocks.
```
What does this mean in practice?

IgWod-IMG replied:
If you look above (lines 1971-1974) there is a piece of code specific to loops that fails if there are block arguments in the merge block, then the next statements (lines 1979-1980) copy arguments from the header block to the merge block. So, running this code for loops would yield arguments that were copied from the header block into the merge block. I'm not sure it's a desired behaviour or at least I don't have / can't think of an example where this is a desired behaviour. So, I don't run this code for loops. Worst case scenario, there will be some uses that escape the loop but that should be caught later. Maybe the comment in the code should make it clearer.

Hope that explains the rational and somehow answers the question.

IgWod committed:
Update comment

IgWod-IMG commented:
I have updated the comment and rebased on top of main to include other SPIR-V patches I merged today - I wanted to make sure it all still builds and passes tests. I'll merge it once all checks passes.

IgWod committed:
Fix formatting
